pydarn/pydmap/io.py

```diff
@@ -0,0 +1,1296 @@
+# Copyright 2018 SuperDARN
+# Authors: Marina Schmidt and Keith Kotyk
+"""
+This file contains classes to reading and writing of DMAP file formats used by
+SuperDARN.
+
+Classes:
+--------
+DmapRead : Reads DMAP files
+DmapWrite : writes DMAP Record structure into a DMAP file
+
+Exceptions:
+-----------
+EmptyFileError
+CursorError
+DmapDataError
+DmapDataTypeError
+ZeroByteError
+MismatchByteError
+ValueError
+
+Future work
+-----------
+logger improvement
+    time performance can improve if we turn off the logger
+DmapWrite cleanup
+    updating and improving DmapWrite
+Date stream reading/writing
+    add the ability to read/write from data streams <-- this may be working but
+                                                        not tested well
+Organization
+    rethink public and private methods? <--- discussion
+
+Notes
+-----
+DmapRead and DmapWrite are updated and improved versions of
+backscatter's pydmap:
+git@github.com:SuperDARNCanada/backscatter.git
+written by Keith Kotyk
+"""
+
+import os
+import struct
+import numpy as np
+import logging
+import collections
+
+from pydarn import pydmap_exceptions
+
+from pydarn import DmapArray
+from pydarn import DmapScalar
+
+# Keeping these global definitions for readability purposes
+# Data types use in DMAP files
+DMAP = 0
+CHAR = 1
+SHORT = 2
+INT = 3
+FLOAT = 4
+DOUBLE = 8
+STRING = 9
+LONG = 10
+UCHAR = 16
+USHORT = 17
+UINT = 18
+ULONG = 19
+
+# Dictionary of DMAP types (key) to quickly convert the format and byte size
+# (value-tuple)
+DMAP_DATA_TYPES = {DMAP: ('', 0),
+                   CHAR: ('c', 1),
+                   SHORT: ('h', 2),
+                   INT: ('i', 4),
+                   FLOAT: ('f', 4),
+                   DOUBLE: ('d', 8),
+                   STRING: ('s', 1),
+                   LONG: ('q', 8),
+                   UCHAR: ('B', 1),
+                   USHORT: ('H', 2),
+                   UINT: ('I', 4),
+                   ULONG: ('Q', 8)}
+
+pydarn_logger = logging.getLogger('pydarn')
+
+class DmapRead():
+    """
+    Reading and testing the integrity of DMAP files/stream.
+    ...
+
+    Attributes:
+    ----------
+    dmap_file : str
+        DMAP file/stream name/data
```

Comment:
I will fix what I can but I will take a more serious look at this when I focus on DmapWrite again since I will go over the documentation again in io.py file to make sure things are updated.